File: 3116-kth-smallest-amount-with-single-denomination-combination/3116-kth-smallest-amount-with-single-denomination-combination.java
class Solution {
    public long gcd(long a, long b) {
        while (b > 0) {
            long c = a;
            a = b;
            b = c % b;
        }
        return a;
    }

    public long findKthSmallest(int[] coins, int k) {
        int n = coins.length;
        long[] LCM = new long[1 << n];
        for (int i = 0; i < n; i++) {
            LCM[1 << i] = coins[i];
        }

        LCM[0] = 1;
        for (int i = 1; i < 1 << n; i++) {
            if (LCM[i] == 0) {
                int q = i & i - 1;
                int r = i & -i;
                LCM[i] = LCM[q] / gcd(LCM[q], LCM[r]) * LCM[r];
            }
        }

        long low = 0, high = 100000000000000L;
        while (high - low > 1) {
            long h = (high + low) / 2;
            long cnt = 0;
            for (int i = 1; i < 1 << n; i++) {
                cnt += h / LCM[i] * (Integer.bitCount(i) % 2 == 0 ? -1 : 1);
            }
            if (cnt >= k) {
                high = h;
            } else {
                low = h;
            }
        }
        return high;
    }
}

File: 3116-kth-smallest-amount-with-single-denomination-combination/3116-kth-smallest-amount-with-single-denomination-combination.cpp
class Solution {
public:
    long long findKthSmallest(vector<int>& coins, int k) {
     
      using ll = long long;
      int n = coins.size();
      vector<ll> mult(1 << n, 1);
      for (int b = 0; b < (1 << n); ++b) {
        for (int i = 0; i < n; ++i) {
          if (b & (1 << i)) {
            mult[b] = mult[b] / gcd(mult[b], coins[i]) * coins[i];
          }
        }
      }
      ll lo = 1, hi = 1e12;
      while (lo < hi) {
        ll mid = (lo + hi) / 2;
        ll cnt = 0;
        for (int b = 1; b < (1 << n); ++b) {
          ll v = mid / mult[b];
          if (__builtin_popcount(b) & 1) cnt += v;
          else cnt -= v;
        }
        if (cnt < k) lo = mid + 1;
        else hi = mid;
      }
      return lo;
        
    }
};